trade-finance-manager-api/src/v1/controllers/auth/auth-provider.js

```diff
@@ -0,0 +1,183 @@
+     * https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#authorizationcoderequest
+     * */
+    response.authCodeUrlRequest = {
+      ...authCodeUrlRequestParams,
+      redirectUri: this.config.redirectUri,
+      responseMode: 'form_post', // recommended for confidential clients
+      codeChallenge: response.pkceCodes.challenge,
+      codeChallengeMethod: response.pkceCodes.challengeMethod,
+    };
+
+    response.authCodeRequest = {
+      ...authCodeRequestParams,
+      redirectUri: this.config.redirectUri,
+      code: '',
+    };
+    response.loginUrl = await msalInstance.getAuthCodeUrl(response.authCodeUrlRequest);
+    return response;
+  }
+
+  /**
+   * handleRedirect original is from Microsoft example, but was modified to meet out minimal needs.
+   * OpendId tokenResponse is enough for us (UKEF), it has all claims (aka fields) required for user data.
+   * We don't save Authorisation code because we don't need Access token.
+   * @param req: Express request object
+   * @param res: Express response object
+   * @param next: Express next function
+   * @returns
+   */
+  async handleRedirect(pkceCodes, origAuthCodeRequest, code, body) {
+    const authCodeRequest = {
+      ...origAuthCodeRequest,
+      code, // authZ code
+      codeVerifier: pkceCodes.verifier, // PKCE Code Verifier
+    };
+
+    const msalInstance = this.getMsalInstance(this.config.msalConfig);
+
+    // TODO: do we need token cache?
+    // msalInstance.getTokenCache().deserialize(req.session.tokenCache);
+    const tokenResponse = await msalInstance.acquireTokenByCode(authCodeRequest, body);
+
+    return tokenResponse.account;
+  }
+
+  redirectAfterLogin(rawState) {
+    const state = JSON.parse(this.cryptoProvider.base64Decode(rawState));
+    return state.redirectTo;
+  }
+
+  /**
+     *
+     * @param req: Express request object
+     * @param res: Express response object
+     * @param next: Express next function
+     */
+  getLogoutUrl() {
+    /**
+     * Construct a logout URI and redirect the user to end the
+     * session with Azure AD. For more information, visit:
+     * https://docs.microsoft.com/azure/active-directory/develop/v2-protocols-oidc#send-a-sign-out-request
+     */
+    return `${this.config.msalConfig.auth.authority}${TENANT_SUBDOMAIN}.onmicrosoft.com/oauth2/v2.0/logout?post_logout_redirect_uri=${POST_LOGOUT_REDIRECT_URI}`;
+  }
+
+  /**
+   * Retrieves oidc metadata from the openid endpoint
+   * @returns
+   */
+  async getAuthorityMetadata() {
+    const endpoint = `${this.config.msalConfig.auth.authority}${TENANT_SUBDOMAIN}.onmicrosoft.com/v2.0/.well-known/openid-configuration`;
+    try {
+      const response = await axios.get(endpoint);
+      return await response.data;
+    } catch (error) {
+      console.error('Error in getAuthorityMetadata', error);
```

Comment:
Should we also throw an error here?

Any maybe add "MSAL" so that when looking at logs it's clear the issue is with MSAL